engineapi/engineapi/middleware.py

```diff
@@ -199,3 +212,157 @@ async def dispatch(
+    """
+    resource_origins_set = set()
+    for resource in origins:
+        origins = resource.resource_data.get("origins", [])
+        for o in origins:
+            try:
+                parse_obj_as(AnyHttpUrl, o)
+                resource_origins_set.add(o)
+            except Exception:
+                logger.info(f"Unable to parse origin: {o} as URL")
+                continue
+
+    return resource_origins_set
+
+
+def check_default_origins(origins: Set[str]) -> Set[str]:
+    """
+    To prevent default origins loss.
+    """
+    for o in ALLOW_ORIGINS:
+        if o not in origins:
+            origins.add(o)
+    return origins
+
+
+def fetch_application_settings_cors_origins(token: str) -> Set[str]:
+    """
+    Fetch application config resources with CORS origins setting.
+    If there are no such resources create new one with default origins from environment variable.
+
+    Should return in any case some list of origins, by default it will be ALLOW_ORIGINS.
+    """
+
+    # Fetch CORS origins configs from resources for specified application
+    resources: BugoutResources
+    try:
+        resources = bc.list_resources(
+            token=token,
+            params={
+                "application_id": MOONSTREAM_APPLICATION_ID,
+                "type": BUGOUT_RESOURCE_TYPE_APPLICATION_CONFIG,
+                "setting": "cors",
+            },
+            timeout=BUGOUT_REQUEST_TIMEOUT_SECONDS,
+        )
+
+    except Exception as err:
+        logger.error(f"Error fetching bugout resources with CORS origins: {str(err)}")
+        return ALLOW_ORIGINS
+
+    # If there are no resources with CORS origins configuration, create new one
+    # with default list of origins from environment variable
+    if len(resources.resources) == 0:
+        try:
+            resource = bc.create_resource(
+                token=MOONSTREAM_ADMIN_ACCESS_TOKEN,
+                application_id=MOONSTREAM_APPLICATION_ID,
+                resource_data={
+                    "type": BUGOUT_RESOURCE_TYPE_APPLICATION_CONFIG,
+                    "setting": "cors",
+                    "user_id": str(MOONSTREAM_ADMIN_USER.id),
+                    "origins": ALLOW_ORIGINS,
+                },
+            )
+            resources.resources.append(resource)
+            logger.info(
+                "Created resource with default CORS origins setting by moonstream admin user"
+            )
+        except Exception as err:
+            logger.error(
+                f"Unable to write default CORS origins to resources: {str(err)}"
+            )
+            return ALLOW_ORIGINS
+
+    resource_origins_set: Set[str] = parse_origins_from_resources(resources.resources)
+    resource_origins_set = check_default_origins(resource_origins_set)
+
+    return list(resource_origins_set)
+
+
+def set_cors_origins_cache(allow_origins: Set[str]) -> None:
+    try:
+        allow_origins_str = ",".join(list(allow_origins))
```

Comment:
Redis is a key-value store. Why aren't we storing the actual origins as keys? It doesn't make sense to me that we are just joining all the origins into a comma-separated string and storing under the `cors` key.

What will happen if there are 1,000,000 origins that we need to whitelist?

I would use a Redis set to store the whitelisted origins: https://redis.io/docs/data-types/sets/

Reply:
Yes, I will use it, I was just following old approach where we were storing CORS origins as string in environment variable.